engine/src/flutter/impeller/geometry/path_builder.cc

```diff
@@ -219,6 +270,120 @@ PathBuilder& PathBuilder::AddRoundRect(RoundRect round_rect) {
   return *this;
 }
 
+PathBuilder& PathBuilder::AddRoundSuperellipse(RoundSuperellipse rse) {
+  if (rse.IsRect()) {
+    return AddRect(rse.GetBounds());
+  }
+
+  auto SuperellipseArcPoints = [](Point* output,
+                                  const RoundSuperellipseParam::Octant& param) {
+    Point start = {param.se_center.x, param.edge_mid.y};
+    const Point& end = param.circle_start;
+    constexpr Point start_tangent = {1, 0};
+    Point circle_start_vector = param.circle_start - param.circle_center;
+    Point end_tangent =
+        Point{-circle_start_vector.y, circle_start_vector.x}.Normalize();
+
+    Scalar factors[2];
+    SuperellipseBezierFactors(factors, param.se_n);
+
+    output[0] = start;
+    output[1] = start + start_tangent * factors[0] * param.se_a;
+    output[2] = end + end_tangent * factors[1] * param.se_a;
+    output[3] = end;
+  };
+
+  auto CircularArcPoints = [](Point* output,
+                              const RoundSuperellipseParam::Octant& param) {
+    Point start_vector = param.circle_start - param.circle_center;
+    Point end_vector =
+        start_vector.Rotate(Radians(-param.circle_max_angle.radians));
+    Point circle_end = param.circle_center + end_vector;
+    Point start_tangent = Point{start_vector.y, -start_vector.x}.Normalize();
+    Point end_tangent = Point{-end_vector.y, end_vector.x}.Normalize();
+    Scalar bezier_factor = std::tan(param.circle_max_angle.radians / 4) * 4 / 3;
+    Scalar radius = start_vector.GetLength();
+
+    output[0] = param.circle_start;
+    output[1] = param.circle_start + start_tangent * bezier_factor * radius;
+    output[2] = circle_end + end_tangent * bezier_factor * radius;
+    output[3] = circle_end;
+  };
+
+  auto AddOctant = [this, &CircularArcPoints, &SuperellipseArcPoints](
+                       const RoundSuperellipseParam::Octant& param,
+                       bool reverse, bool flip,
+                       const Matrix& external_transform) {
+    Matrix transform =
+        external_transform * Matrix::MakeTranslation(param.offset);
+    if (flip) {
+      transform = transform * kFlip;
+    }
+
+    Point circle_points[4];
+    CircularArcPoints(circle_points, param);
```

Comment:
If you have functions that are always returning the same number of points then I think its clearer to return a std::array<Point, 4> rather than write to a raw ptr.